core/src/main/java/hudson/model/Queue.java

```diff
@@ -3009,4 +3010,52 @@ public static Queue getInstance() {
     public static void init(Jenkins h) {
         h.getQueue().load();
     }
+
+    /**
+     * Schedule <tt>Queue.save()</tt> call for near future once items change. Ignore all changes until the time the save
+     * takes place.
+     *
+     * Once queue is restored after a crash, items stages might not be accurate until the next #maintain() - this is not
+     * a problem as the items will be reshuffled first and then scheduled during the next maintainance cycle.
+     *
+     * Implementation note: Queue.load() calls QueueListener hooks for every item deserialized that can hammer the persistance
+     * on load. The problem is avoided by delaying the actual save for the time long enough for queue to load so the save
+     * operations will collapse into one. Also, items are persisted as buildable or blocked in vast majority of cases and
+     * those stages does not trigger the save here.
+     */
+    @Extension
+    @Restricted(NoExternalUse.class)
+    public static final class Saver extends QueueListener implements Runnable {
+        @VisibleForTesting
+        /*package*/ static /*final*/ int DELAY_SECONDS = 60;
```

Comment:
Maybe makes sense to add a system property for managing time

Reply:
'-1' value could be also used to disable this delayed task. It may be feasible if somebody hits a performance degradation. CC @Jimilian who often fixes performance issues on his instance

Reply:
In our particular case we have sometimes ~3k elements in queue and I don't remember when Jenkins crashed last time (JVM crashed or was killed by OOM-killer). So, in our case we want to save the queue state only before graceful restart.

Reply:
What I'm hearing from @Jimilian is maybe we want to support "-1" for disabling autosave.

Reply:
(Reiterating that there's value there.)  Although I think setting it to something like "90000" might be okay, assuming we don't get churn from the listeners.